File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/simpleviewlets/ImageViewlet.kt
package com.crescentflare.piratesgame.components.simpleviewlets

import android.content.Context
import android.content.res.Resources
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.crescentflare.piratesgame.components.utility.CustomNinePatchDrawable
import com.crescentflare.piratesgame.components.utility.CustomThreePatchDrawable
import com.crescentflare.piratesgame.components.utility.ImageSource
import com.crescentflare.piratesgame.components.utility.ViewletUtil
import com.crescentflare.unilayout.views.UniImageView
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil


/**
 * Basic view viewlet: an image view
 */
object ImageViewlet {

    // --
    // Viewlet instance
    // --

    val viewlet: JsonInflatable = object : JsonInflatable {

        override fun create(context: Context): Any {
            return UniImageView(context)
        }

        override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
            if (obj is UniImageView) {
                // Set image
                applyImageSource(obj, ImageSource.fromObject(attributes["source"]))

                // Scale factor
                val scaleType = ScaleType.fromString(mapUtil.optionalString(attributes, "scaleType", ""))
                obj.scaleType = scaleType.toImageViewScaleType()

                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)
                return true
            }
            return false
        }

        override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
            return obj is UniImageView
        }

    }


    // --
    // Helpers
    // --

    fun applyImageSource(imageView: UniImageView?, source: ImageSource?): Boolean {
        if (imageView != null) {
            if (source != null) {
                source.onlineUri?.let {
                    if (source.type == ImageSource.Type.DevServerImage) {
                        val scale = Resources.getSystem().displayMetrics.density / 4f // Place xxxhdpi images in the dev server
                        Glide.with(imageView).load(it).apply(RequestOptions().sizeMultiplier(scale)).into(imageView)
                    } else {
                        Glide.with(imageView).load(it).into(imageView)
                    }
                    if (source.tintColor != 0) {
                        imageView.colorFilter = PorterDuffColorFilter(source.tintColor, PorterDuff.Mode.SRC_IN)
                    } else {
                        imageView.colorFilter = null
                    }
                    return true
                } ?: run {
                    val drawable = source.getDrawable(imageView.context)
                    if (drawable != null) {
                        if (source.threePatch > 0) {
                            imageView.setImageDrawable(prepareThreePatch(drawable, source.threePatch))
                        } else if (source.ninePatch > 0) {
                            imageView.setImageDrawable(prepareNinePatch(drawable, source.ninePatch))
                        } else {
                            imageView.setImageDrawable(drawable)
                        }
                        if (source.tintColor != 0) {
                            imageView.colorFilter = PorterDuffColorFilter(source.tintColor, PorterDuff.Mode.SRC_IN)
                        } else {
                            imageView.colorFilter = null
                        }
                        return true
                    }
                }
            }
            imageView.colorFilter = null
            imageView.setImageResource(0)
        }
        return false
    }

    private fun prepareThreePatch(drawable: Drawable, edgeInset: Int): Drawable? {
        if (drawable is BitmapDrawable) {
            val bitmap = drawable.bitmap
            if (bitmap != null) {
                return CustomThreePatchDrawable(bitmap, edgeInset)
            }
        }
        return null
    }

    private fun prepareNinePatch(drawable: Drawable, edgesInset: Int): Drawable? {
        if (drawable is BitmapDrawable) {
            val bitmap = drawable.bitmap
            if (bitmap != null) {
                return CustomNinePatchDrawable(bitmap, edgesInset, edgesInset)
            }
        }
        return null
    }


    // --
    // Scale type enum
    // --

    enum class ScaleType(val value: String) {

        Center("center"),
        Stretch("stretch"),
        ScaleFit("scaleFit"),
        ScaleCrop("scaleCrop");

        fun toImageViewScaleType(): ImageView.ScaleType {
            return when(this) {
                Stretch -> ImageView.ScaleType.FIT_XY
                ScaleFit -> ImageView.ScaleType.FIT_CENTER
                ScaleCrop -> ImageView.ScaleType.CENTER_CROP
                else -> ImageView.ScaleType.CENTER_INSIDE
            }
        }

        companion object {

            fun fromString(string: String?): ScaleType {
                for (enum in ScaleType.values()) {
                    if (enum.value == string) {
                        return enum
                    }
                }
                return Center
            }

        }

    }

}

File: pirates-ios/PiratesGame/Components/SimpleViewlets/ImageViewlet.swift
//
//  ImageViewlet.swift
//  Basic view viewlet: an image view
//

import UIKit
import UniLayout
import JsonInflator
import AlamofireImage

enum ImageScaleType: String {
    
    case center = "center"
    case stretch = "stretch"
    case scaleFit = "scaleFit"
    case scaleCrop = "scaleCrop"
    
    func toContentMode() -> UIView.ContentMode {
        switch self {
        case .stretch:
            return .scaleToFill
        case .scaleFit:
            return .scaleAspectFit
        case .scaleCrop:
            return .scaleAspectFill
        case .center:
            return .center
        }
    }

}

fileprivate struct AlamofireViewletFilter: ImageFilter {

    private let scale: CGFloat
    private let withRenderingTemplate: Bool
    
    public init(scale: CGFloat = 1, withRenderingTemplate: Bool = false) {
        self.scale = scale
        self.withRenderingTemplate = withRenderingTemplate
    }
    
    public var filter: (Image) -> Image {
        return { image in
            let result = self.scale == 1 ? image : image.af_imageScaled(to: CGSize(width: image.size.width * self.scale, height: image.size.height * self.scale))
            if self.withRenderingTemplate {
                return result.withRenderingMode(.alwaysTemplate)
            }
            return result
        }
    }
    
}

class ImageViewlet {
    
    // --
    // MARK: Viewlet instance
    // --
    
    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return UniImageView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let imageView = object as? UniImageView {
                // Image
                imageView.clipsToBounds = true
                ImageViewlet.applyImageSource(imageView: imageView, source: ImageSource(value: attributes["source"]))
                
                // Scale factor
                let scaleType = ImageScaleType(rawValue: convUtil.asString(value: attributes["scaleType"]) ?? "") ?? .center
                imageView.internalImageView.contentMode = scaleType.toContentMode()
                
                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: imageView, attributes: attributes)
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is UniImageView
        }
        
    }
    
    
    // --
    // MARK: Helper
    // --
    
    @discardableResult
    static func applyImageSource(imageView: UniImageView?, source: ImageSource?) -> Bool {
        return applyImageSource(imageView: imageView?.internalImageView, source: source)
    }
    
    @discardableResult
    static func applyImageSource(imageView: UIImageView?, source: ImageSource?) -> Bool {
        if let onlineUri = source?.onlineUri {
            if let imageUrl = URL(string: onlineUri) {
                let tintColor = source?.tintColor
                let filter = AlamofireViewletFilter(scale: source?.type == .devServerImage ? UIScreen.main.scale / 4 : 1, withRenderingTemplate: tintColor != nil)
                imageView?.af_setImage(withURL: imageUrl, filter: filter)
                imageView?.tintColor = tintColor ?? UIColor.clear
                return true
            }
        } else if var image = source?.getImage() {
            let tintColor = source?.tintColor
            if tintColor != nil {
                image = image.withRenderingMode(.alwaysTemplate)
            }
            if let threePatch = source?.threePatch {
                image = image.stretchableImage(withLeftCapWidth: Int(threePatch), topCapHeight: 0)
            } else if let ninePatch = source?.ninePatch {
                image = image.stretchableImage(withLeftCapWidth: Int(ninePatch), topCapHeight: Int(ninePatch))
            }
            imageView?.tintColor = tintColor ?? UIColor.clear
            imageView?.image = image
            return true
        }
        imageView?.image = nil
        return false
    }
    
}
